h/groups/views.py

```diff
@@ -101,26 +104,28 @@ def _read_group(request, group):
     """
     url = request.route_url('group_read', pubid=group.pubid, slug=group.slug)
 
-    result = search.search(request,
-                           private=False,
-                           params={"group": group.pubid, "limit": 1000})
-    annotations = [presenters.AnnotationHTMLPresenter(
-                       storage.annotation_from_dict(a))
-                   for a in result['rows']]
-
-    # Group the annotations by URI.
-    # Create a dict mapping the (normalized) URIs of the annotated documents
-    # to the most recent annotation of each document.
-    annotations_by_uri = collections.OrderedDict()
-    for annotation in annotations:
-        normalized_uri = uri.normalize(annotation.uri)
-        if normalized_uri not in annotations_by_uri:
-            annotations_by_uri[normalized_uri] = annotation
-            if len(annotations_by_uri) >= 25:
-                break
-
-    document_links = [annotation.document_link
-                      for annotation in annotations_by_uri.values()]
+    # Find documents that have at least one annotation that:
+    # a. is shared, and
+    # b. is in this group.
+    documents = (
```

Comment:
I've tested this locally with a group that has >10k annotations and this query unfortunately takes about 2 minutes on my local machine.
I sat together with @nickstenning and we came up with this:
```
request.db.query(models.Document) \
        .options(joinedload(models.Document.meta)) \
        .options(joinedload(models.Document.document_uris)) \
        .join(models.Document.annotations) \
        .filter(models.Annotation.groupid == group.pubid, models.Annotation.shared == True) \
        .order_by(models.Annotation.updated.desc()) \
        .limit(25).all()
```

This query takes somewhere in between 3-6ms to execute (on the same group). The main difference is that it is using JOIN queries instead lots of subqueries.

The above query orders by `updated` which is not really what we want, but we don't have an index on `created` at the moment. So we should add an index on that column and then order by `created`.